File: reliable_transport/Reliable.c
#include "Reliable.h"

Payload *payloadCreate(uint16_t size, bool fin)
{
    Payload *payload = (Payload *)malloc(sizeof(Payload));
    payload->buf = (char *)malloc(sizeof(char) * size);
    payload->fin = fin;
    return payload;
}

void payloadClose(Payload *payload)
{
    Free(payload->buf);
    Free(payload);
}

void *reliGetPayload(Reliable *reli)
{
    return queueGetUnblock(&reli->buffer);
}

int reliSend(Reliable *reli, Payload *payload)
{
    return queuePut(&reli->buffer, payload, 0);
}

ssize_t reliRecvfrom(Reliable *reli, char *seg_str, size_t size)
{
    return recvfrom(reli->skt, seg_str, size, 0, (struct sockaddr *)&(reli->srvaddr), &(reli->srvlen));
}

ssize_t reliSendto(Reliable *reli, const char *seg_str, const size_t len)
{
    return sendto(reli->skt, seg_str, len, 0, (struct sockaddr *)&reli->srvaddr, reli->srvlen);
}

uint32_t reliUpdateRWND(Reliable *reli, uint32_t _rwnd)
{
    reli->rwnd = _rwnd;
    return reli->rwnd;
}

Timer *reliSetTimer(Reliable *reli, double timesec, void *(*callback)(void *), void *args)
{
    Timer *timer = (Timer *)malloc(sizeof(Timer));
    timerInit(timer, timesec, callback, args);
    heapPush(&reli->timerHeap, timer);
    return timer;
}

static void setSktTimeout(int skt, int timesec)
{
    struct timeval tv;
    tv.tv_sec = timesec;
    tv.tv_usec = 0;
    setsockopt(skt, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
}

static void *reliHandler(void *args);

Reliable *reliCreate(unsigned hport)
{
    Reliable *reli = (Reliable *)malloc(sizeof(Reliable));
    reli->status = CLOSED;
    reli->bytesInFly = 0;
    reli->rwnd = MAX_BDP;
    reli->cwnd = MAX_BDP;
    reli->seg_str = NULL;
    reli->reliImpl = NULL;
    queueInit(&reli->buffer, BUFFER_SIZE);
    heapInit(&reli->timerHeap, timerCmp);

    reli->skt = socket(AF_INET, SOCK_DGRAM, 0);

    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = htons(hport);
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    bind(reli->skt, (struct sockaddr *)&addr, sizeof(addr));

    srand((unsigned)time(NULL));
    return reli;
}

void reliClose(Reliable *reli)
{
    Payload *fin = payloadCreate(0, true);
    reliSend(reli, fin);                 //payload.fin=true;
    pthread_join(reli->thHandler, NULL); //return error number if thHanlder is not initialized

    while (reli->buffer.count > 0)
    {
        Payload *payload = queueFront(&reli->buffer);
        queuePop(&reli->buffer);
        Free(payload->buf);
    }
    queueClear(&reli->buffer);

    while (reli->timerHeap.count > 0)
    {
        Timer *timer = heapTop(&reli->timerHeap);
        heapPop(&reli->timerHeap);
        Free(timer);
    }
    heapClear(&reli->timerHeap);

    reliImplClose(reli->reliImpl);
    Free(reli->seg_str);
    Free(reli);
}

int reliConnect(Reliable *reli, const char *ip, unsigned rport, bool nflag, uint32_t n)
{
    reli->srvaddr.sin_family = AF_INET;
    reli->srvaddr.sin_port = htons(rport);
    reli->srvaddr.sin_addr.s_addr = inet_addr(ip);
    reli->srvlen = sizeof(reli->srvaddr);

    setSktTimeout(reli->skt, 1);
    reli->seg_str = (char *)malloc(sizeof(char) * SEGMENT_SIZE);
    uint32_t seqNum = nflag ? n : rand32();

    int synretry = 0;
    reli->status = SYNSENT;
    while (reli->status != CONNECTED)
    {
        Segment seg = {seqNum, 0, 0, 0, 1, 0, 0, NULL, 0};
        ssize_t len = segPack(&seg, reli->seg_str, SEGMENT_SIZE);
        seg.checksum = reliImplChecksum(reli->seg_str, len);
        len = segPack(&seg, reli->seg_str, SEGMENT_SIZE);
        reliSendto(reli, reli->seg_str, len);
        
        len = reliRecvfrom(reli, reli->seg_str, SEGMENT_SIZE);
        if (len < 0 || reliImplChecksum(reli->seg_str, len) != 0)
        {
            if (synretry > 60)
            {
                reli->status = CLOSED;
                return -1;
            }
            synretry += 1;
            continue;
        }

        segParse(&seg, reli->seg_str, len);
        if (seg.syn && seg.ack && seg.ackNum == (seqNum + 1))
            reli->status = CONNECTED;
    }

    setSktTimeout(reli->skt, 0);
    reli->reliImpl = reliImplCreate(reli, seqNum);
    pthread_create(&(reli->thHandler), NULL, reliHandler, reli);
    return 0;
}

static void *reliHandler(void *args)
{
    Reliable *reli = (Reliable *)args;
    fd_set inputs, outputs;
    while (reli->status != CLOSED)
    {
        FD_ZERO(&inputs);
        FD_SET(reli->skt, &inputs);
        FD_ZERO(&outputs);
        FD_SET(reli->skt, &outputs);
        select(reli->skt + 1, &inputs, &outputs, NULL, NULL);

        if (FD_ISSET(reli->skt, &inputs))
        {
            ssize_t len = reliRecvfrom(reli, reli->seg_str, SEGMENT_SIZE);
            if (len <= 0 || reliImplChecksum(reli->seg_str, len) != 0)
                goto outputLabel;
            Segment seg;
            segParse(&seg, reli->seg_str, len);
            if (reli->status == CONNECTED)
            {
                if (seg.ack && !seg.syn && !seg.fin)
                    reli->bytesInFly -= reliImplRecvAck(reli->reliImpl, &seg, false);
            }
            else if (reli->status == FINWAIT)
            {
                if (seg.ack && !seg.syn && !seg.fin)
                    reli->bytesInFly -= reliImplRecvAck(reli->reliImpl, &seg, false);
                else if (seg.ack && seg.fin)
                {
                    reli->bytesInFly -= reliImplRecvAck(reli->reliImpl, &seg, true);
                    reli->status = CLOSED;
                }
            }
        }

    outputLabel:;
        if (FD_ISSET(reli->skt, &outputs))
        {
            if (reli->status != CONNECTED || reli->bytesInFly >= MIN(reli->rwnd, reli->cwnd))
                goto timerLabel;
            Payload *payload = reliGetPayload(reli);
            if (payload == NULL)
                goto timerLabel;
            if (payload->fin)
            {
                reli->bytesInFly += reliImplSendData(reli->reliImpl, NULL, 0, true);
                reli->status = FINWAIT;
            }
            else
                reli->bytesInFly += reliImplSendData(reli->reliImpl, payload->buf, payload->len, false);
            usleep(1); // Avoid sending too fast and overflowing UDP buffer at the receiver
            payloadClose(payload);
        }

    timerLabel:;
        double now = get_current_time();
        while (reli->timerHeap.count > 0)
        {
            Timer *timer = heapTop(&reli->timerHeap);
            if (now < timer->timestamp)
                break;
            if (timer->enable)
                timerRun(timer);

            heapPop(&reli->timerHeap);
            Free(timer);
        }
    }
    return NULL;
}

File: reliable_transport/Reliable.py
import time
import heapq
import select
import socket
import random
import threading
from ReliableImpl import ReliableImpl
from queue import Queue
from collections import deque
from Util import *


SYNSENT = 0
CONNECTED = 1
FINWAIT = 2
CLOSED = 3


class Reliable:
    def __init__(self, hport=10000):
        self.__dst = (None, None)
        self.__buf = Queue(maxsize=BufferSize)
        self.__skt = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.__skt.bind(('', hport))

        self.status = CLOSED
        self.bytesInFly = 0
        self.rwnd = MaxBDP
        # self.cwnd = 3*BlockSize
        self.cwnd = MaxBDP

        self.timerHeap = []
        self.handlerThread = None
        self.reliImpl = None

    def __del__(self):
        self.__skt.close()

    def getskt(self):
        return self.__skt

    def connect(self, ip='127.0.0.1', rport=7090, n=None):
        self.__dst = (ip, rport)

        self.__skt.settimeout(1)
        if n is None:
            seqNum = random.randint(0, (1 << 32)-1)
        else:
            seqNum = n % SeqNumSpace

        self.status = SYNSENT
        synretry = 0
        while self.status != CONNECTED:
            temp = Segment.pack(seqNum, 0, 0, 0, 1, 0, 0, b'')
            self.sendto(Segment.pack(seqNum, 0, 0, 0, 1, 0, ReliableImpl.checksum(temp), b''))

            try:
                seg_str = self.recvfrom()
            except Exception as e:
                if synretry > 60:
                    self.status = CLOSED
                    return -1
                synretry += 1
                continue
            if ReliableImpl.checksum(seg_str) != 0:
                continue

            seg = Segment(seg_str)
            if seg.syn and seg.ack and seg.ackNum == (seqNum+1) % SeqNumSpace:
                self.status = CONNECTED

        self.__skt.settimeout(None)
        self.reliImpl = ReliableImpl(self, seqNum, seg.seqNum)
        self.handlerThread = Handler(self, self.reliImpl)
        self.handlerThread.start()
        return 0

    def close(self):
        self.send(None)
        if self.handlerThread is not None:
            self.handlerThread.join()

    def getPayload(self):
        return self.__buf.get(block=False)  # raise exception if queue is empty

    def send(self, payload):
        return self.__buf.put(payload)  # block if queue is full

    def recvfrom(self):
        (seg, addr) = self.__skt.recvfrom(SegmentSize)
        return seg

    # Followings are APIs that you may need to use in ReliableImpl
    # sendto: Send a well-formed segment to the destination.
    # 'seg_str' should be an array of bytes (type(seg_str)=<class 'bytes'>) and
    # should not contain UDP header
    def sendto(self, seg_str):
        return self.__skt.sendto(seg_str, self.__dst)

    # updateRWND: Update the receive window size.
    # 'rwnd' means the bytes of the receive window.
    def updateRWND(self, rwnd):
        self.rwnd = rwnd
        return self.rwnd

    # setTimer: Set a timer. Its usage is similar to threading.Timer,
    # but we implement our own Timer in this lab (See Util.py).
    # The function 'callback' will be called with 'args' as arguments
    # after 'timesec' seconds.
    def setTimer(self, timesec, callback, args):
        timer = Timer(timesec, callback, args)
        heapq.heappush(self.timerHeap, timer)
        return timer


class Handler(threading.Thread):
    def __init__(self, reli=None, reliImpl=None):
        super().__init__()
        self.reli = reli
        self.reliImpl = reliImpl

    def run(self):
        inputs = [self.reli.getskt()]
        outputs = [self.reli.getskt()]
        while self.reli.status != CLOSED:
            readable, writable, exceptional = select.select(inputs, outputs, inputs)
            for skt in readable:
                seg_str = self.reli.recvfrom()
                if ReliableImpl.checksum(seg_str) != 0:
                    continue
                seg = Segment(seg_str)
                if self.reli.status == CONNECTED:
                    if seg.ack and not seg.syn and not seg.fin:
                        self.reli.bytesInFly -= self.reliImpl.recvAck(seg, False)
                elif self.reli.status == FINWAIT:
                    if seg.ack and not seg.syn and not seg.fin:
                        self.reli.bytesInFly -= self.reliImpl.recvAck(seg, False)
                    elif seg.ack and seg.fin:
                        self.reli.bytesInFly -= self.reliImpl.recvAck(seg, True)
                        self.reli.status = CLOSED

            for skt in writable:
                if self.reli.status != CONNECTED or self.reli.bytesInFly >= min(self.reli.rwnd, self.reli.cwnd):
                    continue

                try:
                    payload = self.reli.getPayload()
                except Exception as e:
                    continue
                if payload is None:
                    self.reli.bytesInFly += self.reliImpl.sendData(b'', True)
                    self.reli.status = FINWAIT
                else:
                    self.reli.bytesInFly += self.reliImpl.sendData(payload, False)

            timestamp = time.time()
            while len(self.reli.timerHeap) > 0:
                timer = self.reli.timerHeap[0]
                if timestamp < timer.timestamp:
                    break
                heapq.heappop(self.reli.timerHeap)
                if timer.enable:
                    timer.run()
